src/journal-remote/journal-gatewayd.c

```diff
@@ -496,10 +588,16 @@ static int request_handler_entries(
 
         if (m->cursor)
                 r = sd_journal_seek_cursor(m->journal, m->cursor);
+        else if (m->since_set && m->n_skip >= 0)
```

Comment:
Hm? Why `n_skip` needs to be checked??

Reply:
Let's make the truth table of `since_set` / `until_set` / `n_skip >= 0`.
The third column is the expected range of events, using vars :
- $T_0$  : first event in the journal
- $T_{since}$ / $T_{until}$  : specified in since/until range
- $T_{now}$ : last event in the journal

The last column is the branch in the code that should handle this situation : 

Since set | until set | skip | Expected event range | handled by
-- | -- | -- | -- | --
false | false | ≥ 0 | $[T_0 + skip ; T_{now}]$ | `m->n_skip >= 0`
false | false | <0 | $[T_{now} - abs(skip) ; T_{now}]$ | `m->n_skip < 0`
true | false | ≥ 0 | $[T_{since} + skip; T_{now}]$ | `m->since_set && m->n_skip >= 0`
true | false | < 0 | $[T_{now} - abs(skip) ; T_{now}]$* | `m->n_skip < 0`
false | true | ≥ 0 | $[T_0 + skip ; T_{until}]$ | `m->n_skip >= 0`
false | true | < 0 | $[T_{until} - abs(skip); T_{until}]$ | `m->until_set && m->n_skip < 0`
true | true | ≥ 0 | $[T_{since} + skip; T_{until}]$ | `m->since_set && m->n_skip >= 0`
true | true | < 0 | $[T_{until} - abs(skip); T_{until}]$* | `m->until_set && m->n_skip < 0`

<b>*</b> : ... and writing down this table, I realize I made an error on these 2 lines : the range should not start before $T_{since}$, even if `skip` negative argument wants to bring the cursor earlier. I should add a test in `request_reader_entries` to handle this case.

Reply:
> Does it make sense ?

Actually ... I'm not convinced myself by this behaviour. From a user point-of-view, useful queries are (e.g.) :
- events from '2023-10-14 12:00:00'" -> `Range: time=1697284800:`
- at most 10 events from '2023-10-14 12:00:00' to '... 14:00:00 -> `Range: time=1697284800:1697292000::10`
- from 10 events before '2023-10-14 12:00:00' to 40 events after -> naturally expressed as `Range: time=1697284800::-10:50` (or `Range: time=:1697284800:-10:50`)

From the truth table in my previous comment, last query is not possible. Thus, a more sound tuth table is :

Since set | until set | skip | Expected event range | handled by
-- | -- | -- | -- | --
false | false | ≥ 0 | $[T_0 + skip ; T_{now}]$ | `m->n_skip >= 0`
false | false | <0 | $[T_{now} - abs(skip) ; T_{now}]$ | `m->n_skip < 0`
true | false | ≥ 0 | $[T_{since} + skip; T_{now}]$ | `m->since_set && m->n_skip >= 0`
true | false | < 0 | $[T_{since} - abs(skip) ; T_{now}]$* | `m->n_skip < 0`
false | true | ≥ 0 | $[T_0 + skip ; T_{until}]$ | `m->n_skip >= 0`
false | true | < 0 | $[T_{until} - abs(skip); T_{until}]$ | `m->until_set && m->n_skip < 0`
true | true | ≥ 0 | $[T_{since} + skip; T_{until}]$ | `m->since_set && m->n_skip >= 0`
true | true | < 0 | $[T_{since} - abs(skip); T_{until}]$* | `m->until_set && m->n_skip < 0`

<i>*</i> : lines changed from previous truth table.

That's what I've implemented in the last version of the code. Now :
- if we have a `since` value, we set the cursor at this point and move (forward/backwards) by `skip` entries if needed.
- then `since is not set: 
     - if `n_skip >= 0` : we set the cursor at the head
     - else then n_skip < 0 : 
         - if `until` is set : we set the cursor at `until` and move the cursor backwards
         - if not, we position the cursor at journal tail and move the cursor backwards

I've notices an off-by-one error when `skip < 0` in my code that [I fixed](https://github.com/systemd/systemd/pull/29465/files#diff-732cab1fcb807574656bfc1b912fe5e3b41ecba76bdbda6bfb550bf3587f4681R177) in the last commit. I've added [corresponding tests](https://github.com/systemd/systemd/pull/29465/files#diff-cedd829438257913798f459744d6029ed596b21ed43fe69ee806d696384f3343R62).